### DISCUSSION TITLE
Build fails randomly with "There was an error in your GraphQL query"

### DISCUSSION THREAD

**Discussion by leibowitz** [gatsbyjs/gatsby#31444] [Help]
## Description

Build failing only some of the time because of a mismatch between schema definition and graphql query. When analysing the schema with graphql, the type for document is correct. The build even succeeds once in a while. I tried [specifying the type manually](https://www.gatsbyjs.com/docs/reference/graphql-data-layer/schema-customization/#automatic-type-inference) using `createSchemaCustomization` within gatsby-node.js:
```

exports.createSchemaCustomization = ({ actions }) => {
  const { createTypes } = actions;
  const typeDefs = `
    type Frontmatter {
      document: File
    }
  `;
  createTypes(typeDefs);
};
```

But the build still fails once in a while with the same error.

I followed everything from the [documentation](https://www.gatsbyjs.com/docs/how-to/images-and-media/importing-assets-into-files/#querying-for-a-file-in-graphql-using-gatsby-source-filesystem) on how to define files from within markdown content. And as I said before, it works. It's just that the build fails about ~50% of the time.

### Steps to reproduce

Using netlify-cms, define a collection type with a widget file type:
```
collections:
  - name: 'my-type'
    label: 'My type'
    folder: 'src/content/my-folder'
    create: true
    fields:
      - {
          label: 'Template Key'
          name: 'templateKey'
          widget: 'hidden'
          default: 'my-template-type'
        }
      - { label: 'Document', name: 'document', widget: 'file', required: true }
```

And query the data with:
```
  query {
    markdownRemark(
      frontmatter: { templateKey: { eq: "my-template-type" } }
      id: { eq: $id }
    ) {
      frontmatter {
        document {
          name
          publicURL
        }
      }
    }
  }
```

All the code has been put in [this repo](https://github.com/leibowitz/gatsby-starter-netlify-cms-1/tree/repro-31414 ) to make it easier to read

### Expected result

The query should succeed all the time.

### Actual result

Query works on development most of the time (when running gatsby develop), and rarely on production build (when running gatsby build on netlify). I would say it fails about 50% of the time with this error:

> Field "document" must not have a selection since type "String" has no subfields.
> This can happen if you e.g. accidentally added { } to the field "document". If you didn't expect "document" to be of type "String" make sure that your input source and/or plugin is correct.

And it doesn't depend on the content. It just works and fails on the same version, it seems randomly. I'm trying to figure out what I should be looking at, and what could be causing this random error. One of my peer is thinking the concurrency during the build could be causing this. Would that be something to try out? What coud cause non deterministic errors during build, regarding schema inference?

### Environment

>   System:
>     OS: macOS 10.15.4
>     CPU: (8) x64 Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
>     Shell: 3.2.57 - /bin/bash
>   Binaries:
>     Node: 15.3.0 - /usr/local/bin/node
>     Yarn: 1.22.10 - /usr/local/bin/yarn
>     npm: 6.14.11 - /usr/local/bin/npm
>   Languages:
>     Python: 2.7.16 - /usr/bin/python
>   npmPackages:
>     gatsby: ^2.14.2 => 2.19.12 
>     gatsby-background-image: ^0.8.14 => 0.8.19 
>     gatsby-image: ^2.2.9 => 2.2.40 
>     gatsby-plugin-canonical-urls: ^2.1.20 => 2.1.20 
>     gatsby-plugin-create-client-paths: ^2.1.7 => 2.1.22 
>     gatsby-plugin-google-analytics: ^2.1.19 => 2.1.35 
>     gatsby-plugin-lead-forensics: 0.0.2 => 0.0.2 
>     gatsby-plugin-linkedin-insight: ^1.0.1 => 1.0.1 
>     gatsby-plugin-local-search: ^1.1.0 => 1.1.0 
>     gatsby-plugin-manifest: ^2.2.5 => 2.2.41 
>     gatsby-plugin-netlify-cms: ^4.1.7 => 4.1.38 
>     gatsby-plugin-offline: ^2.2.6 => 2.2.6 
>     gatsby-plugin-react-helmet: ^3.1.3 => 3.1.3 
>     gatsby-plugin-robots-txt: ^1.5.0 => 1.5.0 
>     gatsby-plugin-root-import: ^2.0.5 => 2.0.5 
>     gatsby-plugin-sharp: ^2.2.12 => 2.4.5 
>     gatsby-plugin-sitemap: ^2.2.27 => 2.2.27 
>     gatsby-plugin-styled-components: ^3.1.3 => 3.1.19 
>     gatsby-remark-copy-linked-files: ^2.1.9 => 2.1.9 
>     gatsby-remark-images: ^3.1.16 => 3.1.16 
>     gatsby-remark-relative-images: ^0.2.3 => 0.2.3 
>     gatsby-source-filesystem: ^2.1.9 => 2.1.9 
>     gatsby-transformer-remark: ^2.6.14 => 2.6.14 
>     gatsby-transformer-sharp: ^2.2.6 => 2.3.14 
>   npmGlobalPackages:
>     gatsby-cli: 2.8.29
>     gatsby-plugin-lead-forensics: 0.0.2
>

**Comment by LekoArts** [comment]
You're using very outdated packages, e.g. the `gatsby` package is nearly 2 years old. You should at least upgrade to the latest minor release or upgrade to Gatsby v3. As we only maintain Gatsby v2 for critical security issues we won't fix this if it's a bug in v2 and fixed in v3.

**Comment by fishmonster01** [comment]
I've been having the same issue when building on AWS. Tt seemingly usually has to do with querying an image or other file path. Has anyone had any luck solving this issue since then? I'm on the latest Gatsby 4 release.
